src/Hosting/Hosting/src/WebHostBuilderExtensions.cs

```diff
@@ -113,20 +113,23 @@ public static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Action
         return hostBuilder
             .ConfigureServices((context, services) =>
             {
-                services.AddSingleton(typeof(IStartup), sp =>
+                services.AddSingleton(typeof(IStartup), GetStartupInstance);
+
+                [UnconditionalSuppressMessage("Trimmer", "IL2072", Justification = "Startup type created by factory can't be determined statically.")]
+                object GetStartupInstance(IServiceProvider serviceProvider)
                 {
                     var instance = startupFactory(context) ?? throw new InvalidOperationException("The specified factory returned null startup instance.");
 
-                    var hostingEnvironment = sp.GetRequiredService<IHostEnvironment>();
+                    var hostingEnvironment = serviceProvider.GetRequiredService<IHostEnvironment>();
 
                     // Check if the instance implements IStartup before wrapping
                     if (instance is IStartup startup)
                     {
                         return startup;
                     }
 
-                    return new ConventionBasedStartup(StartupLoader.LoadMethods(sp, instance.GetType(), hostingEnvironment.EnvironmentName, instance));
-                });
+                    return new ConventionBasedStartup(StartupLoader.LoadMethods(serviceProvider, instance.GetType(), hostingEnvironment.EnvironmentName, instance));
```

Comment:
Can we change this to use `TStartup` instead of instance.GetType()

Reply:
Discussed more offline and will leave as is